sig-azure/0018-20180711-azure-availability-zones.md

```diff
@@ -0,0 +1,238 @@
+
+PublicIPs will also be created with standard SKU, and they are zone redundant by default.
+
+Note that zonal PublicIPs are not supported. We may add this easily if there’re clear use-cases in the future.
+
+## AzureDisk
+
+When Azure managed disks are created, the `PersistentVolumeLabel` admission controller automatically adds zone labels to them. The scheduler (via `VolumeZonePredicate`) will then ensure that pods that claim a given volume are only placed into the same zone as that volume, as volumes cannot be attached across zones.
+
+> Note that only managed disks are supported. Blob disks don't support availability zones on Azure.
+
+### PVLabeler
+
+`PVLabeler` interface should be implemented for AzureDisk:
+
+```go
+// PVLabeler is an abstract, pluggable interface for fetching labels for volumes
+type PVLabeler interface {
+    GetLabelsForVolume(ctx context.Context, pv *v1.PersistentVolume) (map[string]string, error)
+}
+```
+
+It should return the region and zone of the AzureDisk, e.g.
+
+- `failure-domain.beta.kubernetes.io/region=centralus`
+- `failure-domain.beta.kubernetes.io/zone=centralus-1`
+
+so that the PV will be created with labels:
+
+```sh
+$ kubectl get pv --show-labels
+NAME           CAPACITY   ACCESSMODES   STATUS    CLAIM            REASON    AGE       LABELS
+pv-managed-abc 5Gi        RWO           Bound     default/claim1             46s       failure-domain.beta.kubernetes.io/region=centralus,failure-domain.beta.kubernetes.io/zone=centralus-1
+```
+
+### PersistentVolumeLabel
+
+Besides PVLabeler interface, [PersistentVolumeLabel](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/persistentvolume/label/admission.go) admission controller should also updated with AzureDisk support, so that new PVs could be applied with above labels automatically.
+
+```go
+func (l *persistentVolumeLabel) Admit(a admission.Attributes) (err error) {
+    ...
+    if volume.Spec.AzureDisk != nil {
+        labels, err := l.findAzureDiskLabels(volume)
+        if err != nil {
+            return admission.NewForbidden(a, fmt.Errorf("error querying AzureDisk volume %s: %v", volume.Spec.AzureDisk.DiskName, err))
+        }
+        volumeLabels = labels
+    }
+    ...
+}
+```
+
+### StorageClass
+
+Note that the above interfaces are only applied to AzureDisk PV, not StorageClass. For AzureDisk StorageClass, we should add a new optional parameter `zone` and `zones` (must not be used at the same time) for specifying which zones should be used to provision AzureDisk:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  annotations:
+  labels:
+    kubernetes.io/cluster-service: "true"
+  name: managed-premium
+parameters:
+  kind: Managed
+  storageaccounttype: Premium_LRS
+  zone: "centralus-1"
```

Comment:
The new topology aware dynamic provisioning feature adds a new allowedTopologies field to Storageclass which is intended to replace the zone/zones parameters.

Reply:
Topology-aware dynamic provisioning feature is still alpha in v1.12 if I'm not wrong, so these two parameters are still required before its GA, right?

And thanks of reminding this. `allowedTopologies` feature should also be added here.

Reply:
Correct just wanted to make sure you know that this area is changing and as part of beta, we want to get all the 3 in tree cloud provider plugins to start moving to the new method as well.